### ISSUE TITLE
With using the fp8, after the interruption of training, and then continue , there may be a little difference in loss. Is this caused by the fp8 mechanism?

### ISSUE THREAD

**Issue by zte-tcb** [NVIDIA/TransformerEngine#759]
When I used BF16+Megatron-LM, I interrupted it once and continued training. In this case, the Loss in the first round was exactly the same.  But the use of FP8 is a bit different with the loss. 
For example, at 3,000 rounds, saved weight and optimizer status, then continue training. I'm sure the weight is right loaded, because I set the strict=True when torch.load the 3000 iteration weight. The difference is as follows:

192.169.254.9:  iteration     3001/   10000 | consumed samples:      3073024 | elapsed time per iteration (ms): 1947420.8 | learning rate: 1.361E-04 | global batch size:  1024 | **lm loss: 4.484718E-01** | loss scale: 1.0 | grad norm: 0.076 | number of skipped iterations:   0 | number of nan iterations:   0 |
192.169.254.9:  iteration     3002/   10000 | consumed samples:      3074048 | elapsed time per iteration (ms): 97447.9 | learning rate: 1.361E-04 | global batch size:  1024 | lm loss: 4.572957E-01 | loss scale: 1.0 | grad norm: 0.073 | number of skipped iterations:   0 | number of nan iterations:   0 |

continue traning:
192.169.254.9:  iteration     3001/   10000 | consumed samples:      3073024 | elapsed time per iteration (ms): 173138.9 | learning rate: 1.361E-04 | global batch size:  1024 | **lm loss: 4.484071E-01** | loss scale: 1.0 | grad norm: 0.087 | number of skipped iterations:   0 | number of nan iterations:   0 |
192.169.254.9:  iteration     3002/   10000 | consumed samples:      3074048 | elapsed time per iteration (ms): 97481.4 | learning rate: 1.361E-04 | global batch size:  1024 | lm loss: 4.572799E-01 | loss scale: 1.0 | grad norm: 0.075 | number of skipped iterations:   0 | number of nan iterations:   0 |

**Comment by ptrendx** [comment]
Could you give more details on the cases you are comparing?

Generally speaking there is nothing in the FP8 recipe that would make the result nondeterministic. The typical reason for nondeterminism is FlashAttention, but the fact that you are getting the same results in resuming from BF16 suggests that this is not the problem for you.

Could you do an additional experiment of restarting from the same checkpoint the second time and see if the loss is the same as in point 2 (so the result is different compared to uninterrupted training but deterministic) or a different value?
